File: src/main/java/it/alian/gun/mesmerize/compat/SplashParticle.java
package it.alian.gun.mesmerize.compat;

import it.alian.gun.mesmerize.MConfig;
import it.alian.gun.mesmerize.Mesmerize;
import it.alian.gun.mesmerize.util.Collections;
import org.bukkit.Bukkit;
import org.bukkit.Effect;
import org.bukkit.entity.Entity;
import org.bukkit.entity.ExperienceOrb;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDeathEvent;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public abstract class SplashParticle implements Listener {

    private static SplashParticle impl;

    private static List<Effect> effects;

    @EventHandler(priority = EventPriority.HIGHEST)
    public void onDeath(EntityDeathEvent event) {
        if (!(event.getEntity() instanceof Player) && MConfig.Misc.removeOnDeath) {
            Bukkit.getScheduler().runTaskLater(Mesmerize.instance, () -> event.getEntity().remove(), MConfig.Misc.removeDelay);
            ExperienceOrb orb = event.getEntity().getWorld().spawn(event.getEntity().getLocation(), ExperienceOrb.class);
            orb.setExperience(event.getDroppedExp());
        }
        impl.generateParticles(event.getEntity());
    }

    public abstract void generateParticles(Entity entity);

    public static void init() {
        effects = Arrays.stream(MConfig.Misc.splashParticles).map(Effect::valueOf).collect(Collectors.toList());
        if (MConfig.Misc.enableSplashParticles) {
            try {
                Bukkit.getWorlds().get(0).spigot();
                impl = new PlayEffectImpl();
            } catch (Throwable e) {
                impl = new AbstractImpl();
            }
            Bukkit.getPluginManager().registerEvents(impl, Mesmerize.instance);
        }
    }

    private static class PlayEffectImpl extends SplashParticle {

        @Override
        public void generateParticles(Entity entity) {
            if (entity instanceof LivingEntity && entity.getLastDamageCause() != null && entity.getLastDamageCause() instanceof EntityDamageByEntityEvent
                    && ((EntityDamageByEntityEvent) entity.getLastDamageCause()).getDamager() instanceof Player) {
                Effect effect = Collections.random(effects);
                entity.getWorld().spigot().playEffect(entity.getLocation(), effect, 0, 0,
                        0.5F, 0.5F, 0.5F, 0.5F, MConfig.Misc.particleAmount, MConfig.Misc.particleVisibleRange);
            }
        }
    }

    private static class AbstractImpl extends SplashParticle {

        @Override
        public void generateParticles(Entity entity) {

        }
    }

}

File: src/main/scala/it/alian/gun/mesmerize/compat/SplashParticle.scala
package it.alian.gun.mesmerize.compat

import java.util

import it.alian.gun.mesmerize.MesmerizeDelegate._
import it.alian.gun.mesmerize.scalaapi.Prelude._
import org.bukkit.Effect
import org.bukkit.entity.{Entity, ExperienceOrb, LivingEntity, Player}
import org.bukkit.event.entity.{EntityDamageByEntityEvent, EntityDeathEvent}
import org.bukkit.event.{EventHandler, EventPriority, Listener}

import scala.collection.JavaConverters._

object SplashParticle extends Listener {

  private val impl = if (config("misc.enableSplashParticles", true)) new PlayEffectImpl else new AbstractImpl
  private val effects = config[util.List[String]]("misc.splashParticles").asScala.map(Effect.valueOf).toArray

  @EventHandler(priority = EventPriority.HIGHEST)
  def onDeath(event: EntityDeathEvent): Unit = {
    if (!event.getEntity.isInstanceOf[Player] && config("misc.removeOnDeath")) {
      runTask(config("misc.removeDelay", 0).toLong)(event.getEntity.remove())
      val orb = event.getEntity.getWorld.spawn(event.getEntity.getLocation, classOf[ExperienceOrb])
      orb.setExperience(event.getDroppedExp)
    }
    impl(event.getEntity)
  }

  private class PlayEffectImpl extends (Entity => Unit) {
    override def apply(entity: Entity): Unit = entity match {
      case _: LivingEntity => entity.getLastDamageCause match {
        case event: EntityDamageByEntityEvent => event.getDamager match {
          case _: Player =>
            val effect = effects((math.random() * effects.length).toInt)
            entity.getWorld.spigot.playEffect(entity.getLocation, effect, 0, 0,
              0.5F, 0.5F, 0.5F, 0.5F,
              config("misc.particleAmount", 100), config("misc.particleVisibleRange", 36))
          case _ =>
        }
        case _ =>
      }
      case _ =>
    }
  }

  private class AbstractImpl extends (Entity => Unit) {
    override def apply(entity: Entity): Unit = ()
  }

}
